### PR TITLE
[BEAM-7305] Improve and extend Hazelcast Jet based Java Runner

### PR DESCRIPTION
**Pull request by jbartok**
* added support for Beam UnboundedSources
* optimised how Jet DAG is wired together
* sorted out various metrics related issues
* improved on how JetPipelineResult works
* removed side input related hack
* made cluster member count configurable
* done a bit of cleanup

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by jbartok** [comment]
R: @mxm 
R: @kennknowles

File: runners/jet-experimental/src/main/java/org/apache/beam/runners/jet/JetTransformTranslators.java [hunk 1]
```diff
@@ -79,7 +76,6 @@
     TRANSLATORS.put(PTransformTranslation.FLATTEN_TRANSFORM_URN, new FlattenTranslator());
     TRANSLATORS.put(PTransformTranslation.ASSIGN_WINDOWS_TRANSFORM_URN, new WindowTranslator());
     TRANSLATORS.put(PTransformTranslation.IMPULSE_TRANSFORM_URN, new ImpulseTranslator());
-    TRANSLATORS.put(PTransformTranslation.TEST_STREAM_TRANSFORM_URN, new TestStreamTranslator());
```
Comment (mxm) [inline runners/jet-experimental/src/main/java/org/apache/beam/runners/jet/JetTransformTranslators.java]: Curious, why did you remove this?

File: runners/jet-experimental/src/main/java/org/apache/beam/runners/jet/Utils.java [hunk 2]
```diff
@@ -246,4 +259,34 @@ static boolean usesStateOrTimers(AppliedPTransform<?, ?, ?> appliedTransform) {
...
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      ByteArrayKey that = (ByteArrayKey) o;
+      return Arrays.equals(value, that.value);
+    }
+
+    @Override
+    public int hashCode() {
+      if (hash == 0) {
```
Comment (mxm) [inline runners/jet-experimental/src/main/java/org/apache/beam/runners/jet/Utils.java]: Make `hash` an `Integer` and check for null here?

File: runners/jet-experimental/src/main/java/org/apache/beam/runners/jet/JetTransformTranslators.java (hunk 1 above)
Comment (mxm) [reply runners/jet-experimental/src/main/java/org/apache/beam/runners/jet/JetTransformTranslators.java]: Ah, see that you moved it to the test runner.

**Comment by jbartok** [comment]
Hi @mxm. Yes, I was pondering it yesterday if I should make this pull request out of multiple change-sets or squash them down to a single one... I might not have made the best choice...

The thing is that I'm dumping months of my work into these two change-sets, that's why it looks so non-incremental. The actual development has been done in https://github.com/hazelcast/hazelcast-jet-beam-runner, there are 100+ commits there (debugging has proven simpler if working like this, worth the effort of migrating later). Anyways, from now on pace of development should be slower and I will make it more incremental by issuing more frequent PRs.

As far as reviews are concerned, they have been done to some degree on our module by my Hazelcast colleagues. Here we would need somebody both impartial to Hazelcast and with knowledge of Jet, might be not that simple to find.

**Comment by mxm** [comment]
In the long run it would be nice to have other contributors to the Jet Runner. Of course this can take a bit. I think the first step would be to have the review process with your colleagues here instead of in the fork.

**Comment by jbartok** [comment]
Ok, we'll try to do further development here instead of our own runner repo.
